_resources/port1.0/group/cmake-1.1.tcl

```diff
@@ -34,73 +34,253 @@
+                    destroot.target install/fast
+                    destroot.destdir \
+                                    DESTDIR=${destroot}
+                    # unset the DESTDIR env. variable if it has been set before
+                    destroot.env-delete \
+                                    DESTDIR=${destroot}
+                } else {
+                    ui_error "port:${subport} doesn't support CMake's ${args} generator (don't use cmake.generator on the commandline)"
+                    return -code error "unsupported CMake generator requested (port:${subport})"
+                }
+            }
+            "*Ninja" {
+                if {![tbool cmake.ninja_generator_incompatible]} {
+                    ui_debug "Selecting the Ninja generator ($args)"
+                    depends_build-append \
+                                    port:ninja
+                    build.cmd       ninja
+                    # force Ninja to use the exact number of requested build jobs
+                    build.post_args -j${build.jobs} -v
+                    destroot.target install
+                    # ninja needs the DESTDIR argument in the environment
+                    destroot.destdir
+                    destroot.env-append DESTDIR=${destroot}
+                } else {
+                    ui_error "port:${subport} doesn't support CMake's ${args} generator (don't use cmake.generator on the commandline)"
+                    return -code error "unsupported CMake generator requested (port:${subport})"
+                }
+            }
+            default {
+                ui_msg "generator fallback"
+                if {[llength $args] != 1} {
+                    set msg "cmake.generator requires a single value (not \"${args}\")"
+                } else {
+                    set msg "The \"${args}\" generator is not currently known/supported (cmake.generator is case-sensitive!)"
+                }
+                if {[file tail ${cmake::currentportgroupdir}] eq "group"} {
+                    # we're not being run from the registry so we can raise errors
+                    return -code error ${msg}
+                } else {
+                    ui_error "${msg} (ignoring)"
+                }
+            }
+        }
+    }
+}
 
-pre-configure {
-    file mkdir ${configure.dir}
+default configure.dir {[cmake::build_dir]}
+default build.dir {${configure.dir}}
+default build.post_args {VERBOSE=ON}
+
+# tell CMake to use ccache via the CMAKE_<LANG>_COMPILER_LAUNCHER variable
+# and unset the global configure.ccache option which is not compatible
+# with CMake.
+# See https://stackoverflow.com/questions/1815688/how-to-use-ccache-with-cmake
+proc cmake::ccaching {} {
+    global cmake.ccache prefix
+    if {[tbool cmake.ccache] && [file exists ${prefix}/bin/ccache]} {
+        return [list \
+            -DCMAKE_C_COMPILER_LAUNCHER=${prefix}/bin/ccache \
+            -DCMAKE_CXX_COMPILER_LAUNCHER=${prefix}/bin/ccache]
+    }
```

Comment:
Does this actually work? Unless `configure.ccache` is set, write access to the ccache directory is prohibited by the sandbox. If ccache cannot write to `CCACHE_DIR` it will not error out, but it will also not cache anything, rendering this useless. Also, in trace mode execution of `${prefix}/bin/ccache` will be denied without declaring a dependency on the ccache port.

Support for ccache seems independent from the generators and could be split into a separate commit or be addressed in a subsequent pull request.

Reply:
You mean apart from the fact that it works for me in all my big KF5 builds and that `CMAKE_<LANG>_COMPILER_LAUNCHER` is how one should configure the use of ccache according to what I found online? :) I think I did make the CCACHE_DIR writable for macports_user.
Adding a build dependency on port:ccache is trivial.

I know this PR could be split up because evidently I've not stopped evolving my personal copy waiting for PR #26 to be merged. I'm just not very motivated :-/

Reply:
I do not question that `CMAKE_<LANG>_COMPILER_LAUNCHER` is the correct way for cmake. Making the directory writable for the macports user is not enough as it also needs to be in the sandbox. When ccache encounters a permission problem it will not give you any error, but fail silently and not not write anything to the CCACHE_DIR. Try to empty `/opt/local/var/macports/build/.ccache/` and see if a build actually places any files there.

The dependency on ccache will be [added by base](https://github.com/macports/macports-base/blob/1a78a90fe253ce032822c9881d7c6db4247c2f69/src/port1.0/portutil.tcl#L1453-L1456) when executing the port target, so it will not show up in `port info`. I guess it was done this way as otherwise ccache would show up for almost all ports.

What we would need for this would be a way in base to add the CCACHE_DIR to the sandbox and add the ccache dependency, but not [prepend](https://github.com/macports/macports-base/blob/1a78a90fe253ce032822c9881d7c6db4247c2f69/src/port1.0/portconfigure.tcl#L791-L804) ccache to `CC` in the environment. At the moment this is both controlled by `configure.ccache yes`.